feat(MA): refactor currency handling, integrate it to queries

jabahamondes opened the pull request:
## Problem

We are not using the global currency setting for our "money" calculation. Currently I just implemented on the table query because for trends I will have to add that logic in the query runner.

## Changes

Removed unused enum and consolidated currency options into a new utils file for better organization. Added base currency selection to marketing analytics settings and updated relevant components to utilize the new currency options. Improved type safety in web analytics logic by incorporating base currency into dynamic campaign queries.


## Did you write or update any docs for this change?

- [ ] I've [added or updated the docs](https://posthog.com/handbook/engineering/writing-docs)
- [ ] I've reached out for help from the docs team
- [x] No docs needed for this change

## How did you test this code?

Manually
<img width="962" alt="image" src="https://github.com/user-attachments/assets/2387ae85-b006-4b17-af0e-044cc27e4cd7" />

Commit:
refactor: Update pipelineNodeNew to use source instead of kind

- Changed the parameter from 'kind' to 'source' in urls.pipelineNodeNew across multiple files.
- Updated related logic in sourceWizardLogic and DataWarehouseIntegrationChoice to accommodate the new parameter.
- Ensured consistency in handling external data sources by using lowercase values for source types.

Commit:
feat(marketing): Add Marketing Analytics configuration and settings

- Introduced  type and integrated it into the  interface.
- Added mock data for  in .
- Updated schema to include  and .
- Created new components for Marketing Analytics settings, including  and .
- Implemented logic for managing source mappings in .
- Added routes and UI elements for Marketing Analytics in the data management scene.
- Established backend support with new models and serializers for marketing analytics configuration.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

Commit:
fix: Prs comments

Commit:
Merge branch 'feat/adding-marketing-settings-v2' of github.com:PostHog/posthog into feat/adding-marketing-settings-v2

Commit:
fixing snapshot

Commit:
updated snapshot

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

Commit:
refactor: Update NonNativeExternalDataSourceConfiguration to use dataWarehouseSettingsLogic

- Switched data source logic to utilize dataWarehouseSettingsLogic for fetching dataWarehouseSources.
- Maintained existing functionality by keeping sources_map from marketingAnalyticsSettingsLogic.

Commit:
Merge branch 'feat/adding-marketing-settings-v2' of github.com:PostHog/posthog into feat/adding-marketing-settings-v2

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-marketing-settings-v2

Commit:
fix(migrations): correct dependency for 0763_migrate_team_base_currency and add 0764_teammarketinganalyticsconfig migration

Commit:
fix(migrations): update dependency for 0764_teammarketinganalyticsconfig migration to 0763_migrate_team_base_currency

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-marketing-settings-v2

Commit:
feat(migrations): remove obsolete 0763_migrate_team_base_currency migration and add 0763_teammarketinganalyticsconfig model

Commit:
Purpose mistake

Commit:
fixing

Commit:
refactor(schema): move SourceMap type definition to schema-general and update imports

- Moved `SourceMap` type from `marketing-analytics/utils` to `schema-general.ts` for better organization.
- Updated imports in `webAnalyticsLogic.tsx` and `marketingAnalyticsSettingsLogic.ts` to reflect the new location.
- Adjusted the `_sources_map` field in the migration to ensure it defaults to a non-null value.
- Cleaned up comments in `NonNativeExternalDataSourceConfiguration.tsx` for clarity.

Commit:
feat(marketing-analytics): add self-managed external data source configuration and refactor non-native source handling

- Introduced  for user-managed sources.
- Updated  to utilize shared components for better code reuse.
- Enhanced data mapping logic for marketing analytics, including new URL mapping functions and improved table handling.
- Adjusted schema requirements for clicks and impressions in .
- Added new utility functions for source name mapping based on URL patterns.

Commit:
fix(team): update _sources_map field to enforce non-null constraint

- Changed the _sources_map field in TeamMarketingAnalyticsConfig from null=True to null=False to ensure it always has a value.

Commit:
refactor(data-warehouse): streamline data source URL handling and improve type definitions

- Replaced hardcoded URLs with a base URL constant for better maintainability in `getDataWarehouseSourceUrl`.
- Updated type definitions in `NonNativeExternalDataSourceConfiguration` and `SelfManagedExternalDataSourceConfiguration` for clarity and consistency.
- Adjusted `source_type` in `SharedExternalDataSourceConfiguration` to use a string type for flexibility.
- Changed comment in `removeTableMapping` to reflect the correct behavior of setting field mappings to null instead of undefined.

Commit:
Fixing error when no data in not required

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-marketing-settings-v2

Commit:
Merge branch 'feat/adding-marketing-settings-v2' into feat/adding-s3-and-currency-support-v2

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-s3-and-currency-support-v2

Commit:
feat(web-analytics): update return types to DataWarehouseNode for consistency

Changed return types in webAnalyticsLogic from AnyEntityNode to DataWarehouseNode for improved type safety and clarity in data handling. Updated relevant functions to reflect this change.

Commit:
feat(currency): refactor currency handling and integrate base currency settings

Removed unused enum and consolidated currency options into a new utils file for better organization. Added base currency selection to marketing analytics settings and updated relevant components to utilize the new currency options. Improved type safety in web analytics logic by incorporating base currency into dynamic campaign queries.

Commit:
fix(web-analytics): cast filtered nodes to DataWarehouseNode[]

Added a type assertion to the filter method in webAnalyticsLogic to satisfy the type checker regarding null values. This ensures type safety while maintaining the existing functionality.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
fix(web-analytics): filter required fields in SharedExternalDataSourceConfiguration

Updated the logic in SharedExternalDataSourceConfiguration to filter required fields from MARKETING_ANALYTICS_SCHEMA, ensuring only necessary fields are considered for mapping. This enhances the configuration validation process.

Commit:
feat(web-analytics): integrate marketing analytics logic into web analytics

Added marketing analytics logic to webAnalyticsLogic, enabling the use of valid external tables for dynamic data warehouse node creation. Updated selectors to filter and map external tables, enhancing the overall data handling and configuration process in the web analytics scene. Refactored related components to utilize the new external table structure for improved consistency and clarity.

Commit:
Merge branch 'feat/adding-s3-and-currency-support-v2' of github.com:PostHog/posthog into feat/adding-s3-and-currency-support-v2

Commit:
Merge branch 'master' into feat/adding-s3-and-currency-support-v2

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
refactor(marketing-analytics): simplify external data source configurations

Removed unused  functions from  and . Updated  to directly render the  within the table configuration, enhancing clarity and reducing redundancy. Ensured  in  defaults to an empty object if  is undefined.

Commit:
Merge branch 'feat/adding-s3-and-currency-support-v2' of github.com:PostHog/posthog into feat/adding-s3-and-currency-support-v2

Commit:
feat(marketing-settings): add marketing settings to product tree and settings map

- Introduced a new entry for "Marketing settings" in the product tree metadata with a corresponding feature flag.
- Updated the SETTINGS_MAP to include a new section for marketing settings, replacing the previous external data source configurations with a direct reference to the new MarketingAnalyticsSettings component.
- Adjusted the types to reflect the new marketing settings ID.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
Merge branch 'feat/adding-s3-and-currency-support-v2' into feat/adding-currency-support-to-mka

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-s3-and-currency-support-v2

Commit:
Merge branch 'feat/adding-s3-and-currency-support-v2' of github.com:PostHog/posthog into feat/adding-s3-and-currency-support-v2

Commit:
Merge branch 'feat/adding-s3-and-currency-support-v2' into feat/adding-currency-support-to-mka

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-currency-support-to-mka

Commit:
Merge branch 'master' into feat/adding-currency-support-to-mka

Commit:
feat(currency): add important and other currencies lists for better analytics tracking

Commit:
feat(data-warehouse): introduce DataWarehouseSettingsTab enum and update marketing analytics logic

- Added `DataWarehouseSettingsTab` enum for better type safety in data warehouse settings.
- Updated `external_type` in `ExternalTable` to use the new enum.
- Modified logic to derive `tableType` based on the new enum values.
- Enhanced external tables selector to include `dataWarehouseSources`.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
fix(web-analytics): streamline currency conversion logic in SQL query

- Removed the use of `DEFAULT_CURRENCY` constant and replaced it with a conditional assignment for `costSelect` to enhance clarity and maintainability.
- This change ensures that the currency conversion logic is more concise and adaptable to different base currencies.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
fix(web-analytics): ensure correct currency conversion in SQL query

- Updated `costSelect` to apply `toFloat` after currency conversion, ensuring accurate numeric output.
- This change improves the reliability of cost calculations in web analytics.

Commit:
Merge branch 'feat/adding-currency-support-to-mka' of github.com:PostHog/posthog into feat/adding-currency-support-to-mka

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

jabahamondes commented on frontend/src/lib/components/BaseCurrency/CurrencyDropdown.tsx:
```diff
@@ -1,56 +1,16 @@
 }
 
-// These are the currencies that are most important to show first because they're used by the most customers.
-// Check our web analytics dashboard for the most popular countries from our visitors.
```
All this code was just moved

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
refactor(currency): move important currencies comment to appropriate location

- Moved the comment about important currencies from `utils.tsx` to `currency.ts` for better context and clarity.

Commit:
Merge branch 'feat/adding-currency-support-to-mka' of github.com:PostHog/posthog into feat/adding-currency-support-to-mka

jabahamondes commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/logic/marketingAnalyticsLogic.ts:
```diff
@@ -31,7 +31,7 @@ export const marketingAnalyticsLogic = kea<marketingAnalyticsLogicType>([
             ['sources_map'],
             dataWarehouseSettingsLogic,
-            ['dataWarehouseTables', 'dataWarehouseSourcesLoading'],
+            ['dataWarehouseTables', 'dataWarehouseSourcesLoading', 'dataWarehouseSources'],
```
The source id is necessary for the url creation. This is a bug fix not really related with the pr

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rafaeelaudibert reviewed (approved):
BTW, insights support currency already, I initially used it for revenue analytics but eventually moved to my own query runner.

You can check the `EntityNode` type and see it includes `math_property_revenue_currency`. I was tempted to remove it, but you might wanna use it for now :).

You can look at `revenueAnalyticsLogic` git history to see how I was using it

lricoy reviewed (approved):
+1 to @rafaeelaudibert comments, I knew I've seen related changes before somewhere.

Commit:
feat(web-analytics): enhance currency handling in marketing analytics

- Updated `webAnalyticsLogic` to include `baseCurrency` in `createMarketingDataWarehouseNodes` for improved currency conversion.
- Renamed `base_currency` to `currency` in `MARKETING_ANALYTICS_SCHEMA` for consistency.
- Adjusted `SharedExternalDataSourceConfiguration` to reflect the currency name change.
- Improved timestamp handling in `AggregationOperations` to prevent ClickHouse errors with nullable timestamps.

These changes ensure better currency management and data integrity across the web analytics module.

jabahamondes commented:
@rafaeelaudibert  that's actually great, I updated the aggregation_operations file for making that work. I you can re review that last commit to see if it makes sense to you 🙏

Commit:
Merge remote-tracking branch 'origin/master' into feat/adding-currency-support-to-mka

Commit:
fix(aggregation-operations): handle nullable timestamps in DataWarehouse nodes

- Updated timestamp handling in `AggregationOperations` to use `ifNull` for nullable timestamps, ensuring compatibility with DataWarehouse nodes.
- This change prevents potential errors when timestamps are null, improving data integrity in queries.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
fix(aggregation-operations): refine timestamp handling for DataWarehouse nodes

- Updated the timestamp expression in `AggregationOperations` to use the correct field from `DataWarehouseNode`.
- Enhanced handling of nullable timestamps to prevent ClickHouse errors with `dictGetOrDefault`.
- This change improves data integrity and compatibility in queries involving nullable timestamps.

jabahamondes commented on posthog/hogql_queries/insights/trends/aggregation_operations.py:
```diff
@@ -173,14 +173,21 @@ def _math_func(self, method: str, override_chain: Optional[list[str | int]] = No
+                timestamp_expr = ast.Field(chain=[self.series.timestamp_field])
+            else:
+                # For events, timestamp is the default field
+                timestamp_expr = ast.Field(chain=["timestamp"])
```
Yes, so for DataWarehouseNode it's already required to have that field, so it will always have `timestamp_field`. Now I updated it because the value of the field can be null. On the other `Node types` we default the expr to the previous logic, `timestamp_field` is not used for those nodes.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Commit:
Merge branch 'feat/adding-currency-support-to-mka' of github.com:PostHog/posthog into feat/adding-currency-support-to-mka

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Commit:
fix(aggregation-operations): add timestamp expression for DataWarehouse nodes

Commit:
Merge branch 'feat/adding-currency-support-to-mka' of github.com:PostHog/posthog into feat/adding-currency-support-to-mka

github-actions[bot] committed:
Update query snapshots

Commit:
fix(MarketingAnalyticsSettings): sorting

Commit:
Merge branch 'feat/adding-currency-support-to-mka' of github.com:PostHog/posthog into feat/adding-currency-support-to-mka
